### PR TITLE
Implement URI for pure Wasm (#151)

### PR DESCRIPTION
**Pull request by lostflydev**
Add linkTimeIf branches for all JS-dependent operations in URI:
- URI parsing: js.RegExp.exec() → java.util.regex.Pattern
- Path normalization: js.Array + .join() → Scala Array + StringBuilder
- IPv6 detection: js.RegExp.test() → java.util.regex.Pattern
- Quoting (5 functions): jsReplace(regex, callback) → quoteWithPattern()
- UTF-8 percent-encoding: pure Scala via StandardCharsets.UTF_8

Remove top-level vals with JS types (RegExp, js.Function1) to avoid linker errors; inline them into linkTimeIf JS branches instead.

Unblock URITest in pure Wasm test suite. Add 6 new tests covering multi-component quoting, IPv6, normalize edge cases, and non-ASCII.

### CODE REVIEW

**Comment by sjrd** [comment]
Like on the other PR: Please confirm that [you have not even *looked* at the code of OpenJDK nor its tests](https://github.com/scala-js/scala-js/blob/main/JAVALIB.md), nor any other GPL-licensed code. Likewise, please confirm that your code was not generated by AI (even under your guidance), since in that case there is a 99.9% likelihood that it would have learned from the GPL-licensed code of OpenJDK.

File: javalib/src/main/scala/java/net/URI.scala [hunk 1]
```diff
@@ -437,14 +543,27 @@ object URI {
...
-  private val ipv6Re = new RegExp("^" + ipv6address + "$", "i")
+  private val ipv6ReStr = "^" + ipv6address + "$"
+
+  /** Test whether a host string is an IPv6 address.
+   *  Uses js.RegExp for JS targets and java.util.regex for pure Wasm.
+   */
+  private def testIPv6(host: String): Boolean = {
+    LinkingInfo.linkTimeIf(LinkingInfo.targetPureWasm) {
+      val pattern = java.util.regex.Pattern.compile(ipv6ReStr,
+          java.util.regex.Pattern.CASE_INSENSITIVE)
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: Instead of linktime branching at this level, can you use `RegExpImpl` that is a wrapper for switching between `js.RegExp` and `ju.RegExp`. That will help reduce the diff.

https://github.com/scala-wasm/scala-wasm/blob/7a289dab9212b423237d234bdf6e2a8720119466/javalib/src/main/scala/java/util/regex/RegExpImpl.scala#L17-L45

you can find the usages in `Formatter.scala`, `Float.scala` , and `Double.scala`.

File: javalib/src/main/scala/java/net/URI.scala [hunk 2]
```diff
@@ -298,6 +310,100 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
       new URI(getScheme(), getRawAuthority(), newPath, getQuery(), getFragment())
   }
 
+  /** Pure Wasm implementation of normalize() using Scala Array instead of js.Array. */
+  private def normalizePureWasm(): URI = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: Do we need to re-implement the `normalize` entirely? It seems like we can re-use the same logic by not using jsSplit in pure Wasm path.
I think we can use the same pattern as https://github.com/scala-wasm/scala-wasm/blob/7a289dab9212b423237d234bdf6e2a8720119466/javalib/src/main/scala/java/lang/Float.scala#L340-L387 for abstracting the implementation details of split.

That being said... it seems like we can use `String#split` for both JS and pure Wasm because `parts` is a fixed length array. (If it's a auto growable array, it's worth using js.Array though)

File: javalib/src/main/scala/java/net/URI.scala [hunk 3]
```diff
@@ -765,38 +924,67 @@ object URI {
...
+
   /** matches any character not in unreserved, punct, escaped or other */
-  private val userInfoQuoteRe = new RegExp(
-      // !other = [\u0000-\u00a0\u1680\u2000-\u200a\u202f\u205f\u3000\u2028\u2029]
-      // Char class is: [:!other:^a-z0-9-_.!~*'(),;:$&+=%]
-      "[\u0000- \"#/<>?@\\[-\\^`{-}" +
-      "\u007f-\u00a0\u1680\u2000-\u200a\u202f\u205f\u3000\u2028\u2029]|" +
-      "%(?![0-9a-f]{2})",
-      "ig")
+  private val userInfoQuoteReStr = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: dit, we'd be able to reduce the diff using `RegExpImpl` pattern :)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala [hunk 4]
```diff
@@ -474,4 +474,96 @@ class URITest {
     assertTrue("SSP case-sensitive", new URI("mailto:john") != new URI("mailto:JOHN"))
     assertTrue(new URI("mailto:john") != new URI("MAILTO:jim"))
   }
+
+  // --- Additional tests for pure Wasm compatibility ---
+
+  @Test def multiComponentConstructorQuoting(): Unit = {
```
Comment (tanishiking) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala:482]: Where these tests derived from? Are they added randomly?
Adding tests are fine, but I feel like we'd like to clear about which tests verify what :)

**Review by tanishiking — CHANGES_REQUESTED** [review]
Hey, @lostflydev thank you for working on this. I left some comments.

**Comment by lostflydev** [comment]
> Like on the other PR: Please confirm that [you have not even _looked_ at the code of OpenJDK nor its tests](https://github.com/scala-js/scala-js/blob/main/JAVALIB.md), nor any other GPL-licensed code. Likewise, please confirm that your code was not generated by AI (even under your guidance), since in that case there is a 99.9% likelihood that it would have learned from the GPL-licensed code of OpenJDK.

 I confirm that I have not looked at the code of OpenJDK nor its tests, nor any other GPL-licensed code. The code was not generated by AI.

File: javalib/src/main/scala/java/net/URI.scala (hunk 1 above)
Comment (lostflydev) [reply javalib/src/main/scala/java/net/URI.scala]: Refactored to use RegExpImpl. Added a compile(patternStr, flags) overload to support the case-insensitive flag

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala (hunk 4 above)
Comment (lostflydev) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/net/URITest.scala:482]: Removed tests that were already covered by existing suit, kept only tests for previously untested  
  code paths

File: javalib/src/main/scala/java/net/URI.scala [hunk 5]
```diff
@@ -437,14 +440,23 @@ object URI {
...
+    impl.matches(impl.exec(ipv6RePat, host))
+  }
 
   // URI syntax parser. Based on RFC2396, RFC2732 and adaptations according to
   // JavaDoc.
   // - http://www.ietf.org/rfc/rfc2396.txt (see Appendix A for complete syntax)
   // - http://www.ietf.org/rfc/rfc2732.txt
 
-  private val uriRe = {
+  private val uriReStr = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: why not 

```scala
private val uriRe = {
  ...
  RegExpImpl.impl.compile(uriReStr, "i")
}
```

instead of having `uriReStr` and `uriRePat` ?

It'd be nice to have less diff from the upstream as much as possible. (Actually `RegExpImpl` is introduced for that purpose)

File: javalib/src/main/scala/java/net/URI.scala [hunk 6]
```diff
@@ -437,14 +440,23 @@ object URI {
     // (25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])           # 2001:db8:3:4::192.0.2.33  64:ff9b::192.0.2.33 (IPv4-Embedded IPv6 Address)
   }
 
-  private val ipv6Re = new RegExp("^" + ipv6address + "$", "i")
+  private val ipv6ReStr = "^" + ipv6address + "$"
+
+  private[this] lazy val ipv6RePat =
+    RegExpImpl.impl.compile(ipv6ReStr, "i")
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: Same as `uriRe`, can you directly define `ipv6Re` with `RegExpImpl.impl.compile(...)`

File: javalib/src/main/scala/java/net/URI.scala [hunk 7]
```diff
@@ -437,14 +440,23 @@ object URI {
...
+
+  /** Test whether a host string is an IPv6 address. */
+  private def testIPv6(host: String): Boolean = {
+    import RegExpImpl.impl
+    impl.matches(impl.exec(ipv6RePat, host))
+  }
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: Can you inline this method to reduce diff?

File: javalib/src/main/scala/java/net/URI.scala [hunk 8]
```diff
@@ -217,11 +214,10 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
 
   def normalize(): URI = if (_isOpaque || _path == null) this
   else {
-    import js.JSStringOps._
-
     val origPath = _path
 
-    val segments = origPath.jsSplit("/")
+    // Use String#split which works on both JS and pure Wasm targets
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
```

File: javalib/src/main/scala/java/util/regex/RegExpImpl.scala [hunk 9]
```diff
@@ -29,6 +27,7 @@ private[java] sealed abstract class RegExpImpl {
 
   def compile(patternStr: String): PatRepr
   def compile(patternStr: String, global: Boolean): PatRepr
+  def compile(patternStr: String, flags: String): PatRepr
```
Comment (tanishiking) [inline javalib/src/main/scala/java/util/regex/RegExpImpl.scala]: Instead of receiving JS flag string, what do you think about defining a valid flags for `RegExpImpl` like

```scala
private[java] object RegExpImpl {
 object Flags {
    final val Global = 0x01
    final val CaseInsensitive = 0x02
  }

// and define `compile` as
def compile(patternStr: String, flags: Int)
```

This way, we can define JS/JDK backend as something like

```scala
// js regex
def compile(patternStr: String, flags: Int): PatRepr = {
  val jsFlags =
    (if ((flags & Flags.Global) != 0) "g" else "") +
    (if ((flags & Flags.CaseInsensitive) != 0) "i" else "")
  new js.RegExp(patternStr, jsFlags)

// JDK
  var patternFlags = 0x00
  if ((flags & Flags.CaseInsensitive) != 0)
    patternFlags |= Pattern.CASE_INSENSITIVE
  Pattern.compile(patternStr, patternFlags)
```

WDYT?

File: project/Build.scala [hunk 10]
```diff
@@ -2364,10 +2360,10 @@ object Build {
 
                 // javalib/util
                 !endsWith(f, "/DateTest.scala") && // js.Date
-                !endsWith(f, "/PropertiesTest.scala") && // Date.toString
+                !endsWith(f, "/PropertiesTest.scala") // Date.toString
 
                 // javalib/net
-                !endsWith(f, "/net/URITest.scala") // URI.normalize
+                // (URITest is now compatible with pure Wasm)
```
Comment (tanishiking) [inline project/Build.scala]:
```suggestion
```

File: project/Build.scala [hunk 11]
```diff
@@ -2364,10 +2360,10 @@ object Build {
 
                 // javalib/util
                 !endsWith(f, "/DateTest.scala") && // js.Date
-                !endsWith(f, "/PropertiesTest.scala") && // Date.toString
+                !endsWith(f, "/PropertiesTest.scala") // Date.toString
 
                 // javalib/net
```
Comment (tanishiking) [inline project/Build.scala]:
```suggestion
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 12]
```diff
@@ -806,48 +861,51 @@ object URI {
...
    */
-  private def quoteAuthority(str: String) = {
-    import js.JSStringOps._
-    str.jsReplace(authorityQuoteRe, quoteStr)
-  }
+  private def quoteAuthority(str: String): String =
+    quoteReplace(str, authorityQuotePat)
 
   /** matches any character not in unreserved, reserved, escaped or other */
-  private val illegalQuoteRe = new RegExp(
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala:826]: Why don't we just

```scala
  private val illegalQuoteRe = Pattern.compile(
```
There's no `illegalQuoteReStr` usage elsewhere

File: javalib/src/main/scala/java/net/URI.scala [hunk 13]
```diff
@@ -279,17 +275,24 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
+      if (prependDot)
+        sb.append(".")
+      var i = 0
+      while (i < outIdx) {
+        if (i != 0 || prependDot) sb.append("/")
+        sb.append(segments(i))
+        i += 1
+      }
+      sb.toString
+    }
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: What do you think about something like this? It's basically same thing you do, but no copies, and we can use `scalaOps.mkString` instead of writing loop by hand :) 

```scala
    val prependDot = outIdx != 0 && segments(0).contains(":")
    val normalized =
      ju.Arrays.asList(segments).subList(0, outIdx).scalaOps.mkString("", "/", "")

    val newPath =
      if (prependDot) "./" + normalized
      else normalized
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 14]
```diff
@@ -32,14 +29,14 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
    *  This is a local val for the primary constructor. It is a val,
    *  since we'll set it to null after initializing all fields.
    */
-  private[this] var _fld: RegExp.ExecResult = URI.uriRe.exec(origStr)
-  if (_fld == null)
+  private[this] var _parsed: Array[String] = URI.parseURI(origStr)
+  if (_parsed == null)
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private[this] var _fld = RegExpImpl.impl.exec(URI.uriRe, origStr)
  if (!RegExpImpl.impl.matches(_fld))
```

?

File: javalib/src/main/scala/java/net/URI.scala [hunk 15]
```diff
@@ -32,14 +29,14 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
     throw new URISyntaxException(origStr, "Malformed URI")
 
-  private val _isAbsolute = undefOrIsDefined(_fld(AbsScheme))
-  private val _isOpaque = undefOrIsDefined(_fld(AbsOpaquePart))
+  private val _isAbsolute = _parsed(AbsScheme) != null
+  private val _isOpaque = _parsed(AbsOpaquePart) != null
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private val _isAbsolute = RegExpImpl.impl.exists(_fld, AbsScheme)
  private val _isOpaque = RegExpImpl.impl.exists(_fld, AbsOpaquePart)
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 16]
```diff
@@ -32,14 +29,14 @@ final class URI(origStr: String) extends Serializable with Comparable[URI] {
...
 
-  @inline private def fld(idx: Int): String = undefOrGetOrNull(_fld(idx))
+  @inline private def fld(idx: Int): String = _parsed(idx)
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  @inline private def fld(idx: Int): String = RegExpImpl.impl.getOrElse(_fld, idx, null)
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 17]
```diff
@@ -584,9 +596,30 @@ object URI {
...
+  }
+
+  private[this] lazy val uriRePat =
+    RegExpImpl.impl.compile(uriReStr, "i")
 
-    new RegExp(uriRef, "i")
+  /** Parse a URI string, returning an Array[String] of matched groups
+   *  (null for non-matching groups), or null if the string is not a valid URI.
+   */
+  private def parseURI(str: String): Array[String] = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: If we do like `RegExpImpl.impl.exec(URI.uriRe, origStr)`, I think we don't need this method anymore.

File: javalib/src/main/scala/java/net/URI.scala [hunk 18]
```diff
@@ -765,39 +799,60 @@ object URI {
     res
   }
 
+  /** Replace all matches of a compiled regex pattern with percent-encoded form. */
+  private def quoteReplace(str: String,
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: Can we define a `def replaceAll(pattern: PatRepr, string: String, replacer: Function[String, String]): String` in `RegExpImpl`?

We should be able to define JS/Java impl with:

```scala
def replaceAll(pattern: PatRepr, string: String, replacer: Function[String, String]): String = {
  if (!pattern.global)
    throw new IllegalArgumentException("replaceAll requires a global pattern")
  val jsFunc: js.Function1[String, String] = { (matched: String) =>
    replacer.apply(matched)
  }
  string.jsReplace(pattern, jsFunc)
}


// Java
def replaceAll(pattern: PatRepr, string: String, replacer: Function[String, String]): String = {
  pattern.matcher(string).replaceAll(new Function[MatchResult, String] {
    def apply(result: MatchResult): String = replacer.apply(result.group())
  })
}
```

**Review by tanishiking — CHANGES_REQUESTED** [review]
Thanks, overall it looks good, I think we can simplify it :)

File: javalib/src/main/scala/java/util/regex/RegExpImpl.scala [hunk 19]
```diff
@@ -83,6 +100,22 @@ private[java] object RegExpImpl {
       if (r == null) -1
       else pattern.lastIndex
     }
+
+    def replaceAll(pattern: PatRepr, string: String, mapper: StringMapper): String = {
+      val sb = new java.lang.StringBuilder()
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/regex/RegExpImpl.scala]: Drive-by comment: using a `StringBuilder` is counter-productive in Scala.js and Scala.js-on-Wasm. Use a local `var resultString: String = ""` append to it with `+=`.

File: javalib/src/main/scala/java/net/URI.scala [hunk 20]
```diff
@@ -765,39 +760,45 @@ object URI {
     res
   }
 
+  private object QuoteStrMapper extends java.util.function.Function[String, String] {
+    def apply(t: String): String = quoteStrFn(t)
+  }
+
   /** matches any character not in unreserved, punct, escaped or other */
-  private val userInfoQuoteRe = new RegExp(
+  private[this] lazy val userInfoQuotePat = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private val userInfoQuoteRe = RegExpImpl.impl.compile(
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 21]
```diff
@@ -12,14 +12,13 @@
...
 
-import scala.scalajs.js.RegExp
-import scala.scalajs.js
-
 import scala.annotation.tailrec
 
-import java.lang.Utils._
 import java.nio._
 import java.nio.charset.{CodingErrorAction, StandardCharsets}
+import java.util.{Arrays => juArrays}
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: codingstyle

```suggestion
import java.{util => ju}
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 22]
```diff
@@ -765,39 +760,45 @@ object URI {
     res
   }
 
+  private object QuoteStrMapper extends java.util.function.Function[String, String] {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]: codingstyle:

```suggestion
  private object QuoteStrMapper extends ju.function.Function[String, String] {
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 23]
```diff
@@ -765,39 +760,45 @@ object URI {
...
-    str.jsReplace(userInfoQuoteRe, quoteStr)
-  }
+  private def quoteUserInfo(str: String): String =
+    RegExpImpl.impl.replaceAll(userInfoQuotePat, str, QuoteStrMapper)
 
   /** matches any character not in unreserved, punct, escaped, other or equal
    *  to '/' or '@'
    */
-  private val pathQuoteRe = new RegExp(
+  private[this] lazy val pathQuotePat = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private val pathQuoteRe = RegExpImpl.impl.compile(
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 24]
```diff
@@ -806,48 +807,49 @@ object URI {
    *  in IPv6 addresses, but technically speaking they are in reserved
    *  due to RFC2732).
    */
-  private val authorityQuoteRe = new RegExp(
+  private[this] lazy val authorityQuotePat = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private[this] lazy val authorityQuoteRe = RegExpImpl.impl.compile(
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 25]
```diff
@@ -806,48 +807,49 @@ object URI {
...
-  private def quoteAuthority(str: String) = {
-    import js.JSStringOps._
-    str.jsReplace(authorityQuoteRe, quoteStr)
-  }
+  private def quoteAuthority(str: String): String =
+    RegExpImpl.impl.replaceAll(authorityQuotePat, str, QuoteStrMapper)
 
   /** matches any character not in unreserved, reserved, escaped or other */
-  private val illegalQuoteRe = new RegExp(
+  private[this] lazy val illegalQuotePat = {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private val illegalQuoteRe = RegExp.impl.compile(
```

File: javalib/src/main/scala/java/net/URI.scala [hunk 26]
```diff
@@ -806,48 +807,49 @@ object URI {
...
+  private def quoteIllegal(str: String): String =
+    RegExpImpl.impl.replaceAll(illegalQuotePat, str, QuoteStrMapper)
 
   /** matches characters not in ASCII
    *
    *  Note: It is important that the match is maximal, since we might encounter
    *  surrogates that need to be encoded in one shot.
    */
-  private val nonASCIIQuoteRe = new RegExp("[^\u0000-\u007F]+", "g")
+  private[this] lazy val nonASCIIQuotePat =
```
Comment (tanishiking) [inline javalib/src/main/scala/java/net/URI.scala]:
```suggestion
  private val nonASCIIQuoteRe = RegExpImpl.impl.compile("[^\u0000-\u007F]+", RegExpImpl.Flags.Global)
```

File: javalib/src/main/scala/java/util/regex/RegExpImpl.scala [hunk 27]
```diff
@@ -37,11 +38,21 @@ private[java] sealed abstract class RegExpImpl {
...
 
 private[java] object RegExpImpl {
-  val impl = LinkingInfo.linkTimeIf[RegExpImpl](
-      moduleKind == MinimalWasmModule || moduleKind == WasmComponent) {
+  object Flags {
+    final val Global = 0x01
+    final val CaseInsensitive = 0x02
+  }
+
+  val impl = linkTimeIf[RegExpImpl](targetPureWasm) {
```
Comment (tanishiking) [inline javalib/src/main/scala/java/util/regex/RegExpImpl.scala]:
```suggestion
  val impl = LinkingInfo.linkTimeIf[RegExpImpl](
      moduleKind == MinimalWasmModule || moduleKind == WasmComponent) {
```

**Review by tanishiking — CHANGES_REQUESTED** [review]
Thanks @lostflydev almost LGTM 🙌 . Left several coding style comments, and we have to fix the compilation error in `RegExpImpl.scala`.

Also, could you rebase commits into a single commit?

**Review by tanishiking — APPROVED** [review]
LGTM, thanks a lot @lostflydev !
